File: Tree/House_Robber_III.py
# Problem link :- https://leetcode.com/problems/house-robber-iii/

"""The thief has found himself a new place for his thievery again. There is only one 
entrance to this area, called the "root." Besides the root, each house has one and only 
one parent house. After a tour, the smart thief realized that "all houses in this place 
forms a binary tree". It will automatically contact the police if two directly-linked 
houses were broken into on the same night.Determine the maximum amount of money the 
thief can rob tonight without alerting the police."""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rob(self, root: TreeNode) -> int:
        def helper(root):
            if not root:
                return (0,0)
            left = helper(root.left)
            right = helper(root.right)
            rob = root.val + left[1] + right[1]
            not_rob = max(left) + max(right)
            
            return [rob,not_rob]
        return max(helper(root))

File: Tree/HouseRobberIII.cpp
//Problem Link:- https://leetcode.com/problems/house-robber-iii/

#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    pair<int,int> helper(TreeNode* root){
        if (!root){
            return {0,0};
        }
        pair<int,int> leftans = helper(root->left);
        pair<int,int>rightans = helper(root->right);
        int rob = root->val + leftans.second + rightans.second;
        int not_rob = max(leftans.first,leftans.second) + max(rightans.first,rightans.second);
        return {rob,not_rob};
    }
    
    int rob(TreeNode* root) {
        if (root == NULL){
            return 0;
        }
        if (!root->left && !root->right){
            return root->val;
        }
        pair<int,int> result = helper(root);
        return max(result.first,result.second);
    }
};